wallet-next/src/bin/pwalletd.rs

```diff
@@ -1,22 +1,210 @@
+
+            let pool = SqlitePool::connect(&env::var("DATABASE_URL")?).await?;
+            let storage = Storage::new(pool);
+            storage.migrate().await?;
+
+            // TODO: select a custody service here to provide the wallet data source and update local sqlite storage as needed
+
+            // Fetch chain params if necessary so we can .expect() on them.
+            if storage.chain_params().await.is_none() {
+                chain_params(&client, &mut storage).await?;
+            }
+
+            // Always sync pwalletd on startup.
+            sync(&client, &mut storage).await?;
+            // Retrieve asset list
+            assets(&client, &mut storage).await?;
+
+            // TODO: this is just a sqlite usage stub
+
+            let row = storage.insert_table().await?;
+            let x = storage.read_table().await?;
+
+            println!(
+                "Hello, pwalletd! I got stuff from sqlite: row {} value {}",
+                row, x
+            );
+
+            // TODO: start gRPC service and respond to command requests
+            // let wallet_server = tokio::spawn(
+            //     Server::builder()
+            //         .trace_fn(|req| match remote_addr(req) {
+            //             Some(remote_addr) => tracing::error_span!("wallet_query", ?remote_addr),
+            //             None => tracing::error_span!("wallet_query"),
+            //         })
+            //         .add_service(WalletServer::new(storage.clone()))
+            //         .serve(
+            //             format!("{}:{}", host, wallet_query_port)
+            //                 .parse()
+            //                 .expect("this is a valid address"),
+            //         ),
+            // );
+        }
+    }
+
+    Ok(())
+}
+
+pub async fn sync(client: &ObliviousQueryClient<Channel>, storage: &mut Storage) -> Result<()> {
+    tracing::info!("starting client sync");
+
+    let start_height = storage
+        .last_block_height()
+        .await
+        .map(|h| h + 1)
+        .unwrap_or(0);
+    let mut stream = client
+        .compact_block_range(tonic::Request::new(CompactBlockRangeRequest {
+            start_height,
+            end_height: 0,
+            chain_id: storage
+                .chain_id()
+                .await
+                .ok_or_else(|| anyhow::anyhow!("missing chain_id"))?,
+        }))
+        .await?
+        .into_inner();
+
+    let mut count = 0;
+    while let Some(block) = stream.message().await? {
+        storage.scan_block(block.try_into()?)?;
+        // very basic form of intermediate checkpointing
+        count += 1;
+        if count % 1000 == 1 {
+            storage.commit()?;
+            tracing::info!(height = ?storage.last_block_height().await.unwrap(), "syncing...");
+        }
+    }
+
+    storage.prune_timeouts();
+    storage.commit()?;
+    tracing::info!(end_height = ?storage.last_block_height().await.unwrap(), "finished sync");
+    Ok(())
+}
```

Comment:
This method should be deleted, since it's copied from `pcli`, but its state model here is different, and this code doesn't fit (we don't want a bare method on a `Storage` object, we need a way to instantiate a service that does the syncing).